### DISCUSSION TITLE
How to specify a fallback package for unpackaged files?

### DISCUSSION THREAD

**Discussion by daandemeyer** [rpm-software-management/rpm#2907] [Q&A] [answered]
For hacking on systemd, I'm looking into building rpms from upstream sources. Because we're building from upstream sources, there are naturally unpackaged files depending on how the %files directives are handled. I would like to assign any unpackaged upstream files that haven't been explicitly assigned to a package yet to the systemd package. Is there any way I can specify to rpm that any unpackaged files should be assigned to a specific package?

**Comment by pmatilai** [comment]
No.

systemd is widely packaged already, I'd pick up an existing spec as a basis instead of reinventing the wheel.

**Reply by daandemeyer** [reply]
@pmatilai That's exactly what I'm doing. I'm trying to reuse the opensuse spec. The same problem still applies because this spec is written against a specific systemd version so when upstream introduces new files they aren't assigned to any package yet by the spec.

**Reply by pmatilai** [reply]
Right. Well, that's exactly why the *packager* needs to decide where those files belong.

**Reply by daandemeyer** [reply]
@pmatilai Yes but these packaging specs are downstream. You end up with a chicken and egg problem. If we want to use the downstream packaging specs for building upstream testing packages, the downstream spec needs to be updated before any upstream PR introducing new unpackaged files can be tested and merged. But the downstream spec can only be updated once the upstream PR is merged, because otherwise it would fail to build because the newly added files are not yet installed so you get into a circular dependency.

One way out of this circular dependency is to be able to specify that, when building packages from upstream sources, any unpackaged files go into a fallback package (which in systemd's case would be the main systemd package).

**Reply by pmatilai** [reply]
That's a workflow problem really. Making rpm look the other way is not a solution.

**Reply by pmatilai** [reply]
That said, you can already make rpm ignore unpackaged files. Just flip `%_unpackaged_files_terminate_build` to zero.

**Reply by daandemeyer** [reply]
@pmatilai I found that flag, but that's not a solution, we want these new files to actually get packaged so we can install the development with those new files so we can run tests with them.

Anyway, the Fedora spec already handles this by dynamically generating the files lists so I'll probably try to work the suse spec maintainers to do the same thing.

**Reply by daandemeyer** [reply, marked as answer]
So what I ended up doing is something like the following:

```sh
build() {
    # TODO: Replace meson_build and meson_install overrides with "--undefine __meson_verbose" once
    # https://github.com/mesonbuild/meson/pull/12835 is available.
    rpmbuild \
        -bb \
        --build-in-place \
        --define "_topdir /var/tmp" \
        --define "_sourcedir rpm" \
        --define "_rpmdir $PACKAGEDIR" \
        --define "__check_files sh -c '$(rpm --eval %__check_files) | tee /tmp/unpackaged-files'" \
        "$@" \
        rpm/systemd.spec
}

if ! build; then
    EXIT_STATUS=$?
    if [ ! -s /tmp/unpackaged-files ]; then
        exit $EXIT_STATUS
    fi

    # rpm will append to any existing systemd.lang so delete it explicitly so we don't get duplicate file
    # warnings.
    rm systemd.lang

    cat /tmp/unpackaged-files >>rpm/files.systemd
    build --noprep --nocheck
fi
```

Basically we run `rpmbuild` once and if it fails, we check if there are any unpackaged files. If there are unpackaged files, we append those to the files.systemd file which lists the files of the systemd package and then run `rpmbuild` again with `--noprep` and `--nocheck`.

To get the unpackaged files, we override the `__check_files` macro and have it write its output to `/tmp/unpackaged-files`  in addition to writing it to the standard output.

**Reply by lnussel** [reply]
interesting hack. Would be nice if rpm could do that natively indeed. Would help a lot to build git head of software based on the spec file of a stable version. Might actually be doable by breaking the loop in `processBinaryFiles()` into a two stage process.
